Спринт 17

ReyyL opened the pull request:
В последних коммитах добавил немного контекста по замечаниям

Commit:
Добавил фильтры, темную тему, локализацию и аналитику

Commit:
Добавил аналитику по фильтрам

Commit:
Добавил скриншот тест

Commit:
Добавил контекстное меню

Commit:
Добавил поиск и экран статистики

andrew-kuzn-yndx commented on Tracker/CoreData/TrackerRecordStore.swift:
```diff
@@ -75,11 +61,14 @@ final class TrackerRecordStore: NSObject {
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return [] }
+        let managedContext = appDelegate.persistentContainer.viewContext
+        let fetchRequest = NSFetchRequest<TrackerRecordCoreData>(entityName: "TrackerRecordCoreData")
+        let trackerRecordCoreDataArray = try! managedContext.fetch(fetchRequest)
```
В таких местах с try! может возникать проблема
ЧТо-то пойдет не так и приложение словит краш
Поэтому лучше использовать использовать do-catch конструкцию, чтобы обработать разные сценарии

andrew-kuzn-yndx commented on Tracker/NewTrackers/TrackerCreationViewController.swift:
```diff
@@ -17,27 +17,21 @@ class TrackerCreationViewController: UIViewController {
         collectionView.allowsMultipleSelection = true
+        collectionView.isScrollEnabled = false
         collectionView.register(EmojiCollectionViewCell.self, forCellWithReuseIdentifier: "cell1")
         collectionView.register(ColorCollectionViewCell.self, forCellWithReuseIdentifier: "cell2")
```
Старайся не указывать такие названия "cell2"
Лучше сделать статическое свойство внутри ячейки, пусть reuseIdentifier хранится там, например:
static let reuseIdentifier = "ColorCollectionViewCell"

и вызывать тут как: ColorCollectionViewCell.reuseIdentifier

andrew-kuzn-yndx commented on Tracker/TrackerViewController/FilterController.swift:
```diff
@@ -0,0 +1,114 @@
+        if indexPath == selectedIndex {
+            cell.accessoryType = .checkmark
+        } else {
+            cell.accessoryType = .none
```
Можно сделать более элегантно:
let type = indexPath == selectedIndex ? .checkmark : .none
cell.accessoryType = type

Commit:
Пофиксил баги
PS. 9 баг - не баг, в требованиях указано что трекер нельзя отметить в будущих числах

Commit:
Еще фиксы
"нет код ревью от коллег", у нас нет распределения на команды, оно только в следующем спринте будет

andrew-kuzn-yndx commented on Tracker/NewTrackers/TrackerCreationViewController.swift:
```diff
@@ -234,7 +234,7 @@ extension TrackerCreationViewController: UICollectionViewDataSource {
             
         } else {
-            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell2", for: indexPath) as? ColorCollectionViewCell else {
+            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "colorCell", for: indexPath) as? ColorCollectionViewCell else {
```
Ты ведь добавил статическое свойство в ячейку) Видимо просто забыл тут поправить
ColorCollectionViewCell.reuseColorIdentifier

andrew-kuzn-yndx commented on Tracker/TrackerViewController/FilterController.swift:
```diff
@@ -15,6 +15,8 @@ final class FilterController: UIViewController {
     
     var selectedIndex: IndexPath?
     
+    var currentFilter = "Все трекеры"
```
Давай это и выше тоже сделаем приватными?

andrew-kuzn-yndx commented on Tracker/TrackerViewController/TrackerViewController.swift:
```diff
@@ -648,20 +658,26 @@ extension TrackerViewController: UISearchResultsUpdating {
                 collectionView.isHidden = true
+            } else {
+                notFoundView.isHidden = true
+                collectionView.isHidden = false
```
Тут можно упростить:
notFoundView.isHidden = !result.isEmpty
collectionView.isHidden = result.isEmpty

Commit:
minor fixes
